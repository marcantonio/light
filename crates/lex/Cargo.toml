[package]
name = "lex"
version = "0.0.0"
edition = "2021"

[dependencies]
itertools = "0.10.5"
serde = { version = "1.0.136", features = ["derive"] }

common = { path = "../common" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }
