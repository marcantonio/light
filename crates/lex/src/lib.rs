use serde::Serialize;
use std::iter::Peekable;

use common::Operator;
pub use token::{Token, TokenType};

#[cfg(test)]
mod tests;
pub mod token;

pub type LexResult = std::result::Result<Token, LexError>;

pub struct Lex {
    stream: Peekable<StreamIter<char>>,
    pub tokens: Vec<Token>,
}

impl Lex {
    pub fn new(input: &str) -> Self {
        Lex { stream: StreamIter::new(input).peekable(), tokens: vec![] }
    }

    // Scan all input
    pub fn scan(mut self) -> Result<Vec<Token>, LexError> {
        loop {
            let token = self.lex()?;
            if token.is_eof() {
                break;
            }
            self.tokens.push(token);
        }
        Ok(self.tokens)
    }

    // Recursively process enough characters to produce one token
    fn lex(&mut self) -> LexResult {
        use TokenType::*;

        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => unreachable!("can't lex nothing"),
        };

        // Inject a semicolon if certain tokens occur at the end of the line or
        // EOF. If EOF, make sure the context is right.
        if cur.value == '\n' && self.should_add_semicolon() {
            return Ok(Token::new(Semicolon(true), cur.line, cur.column));
        } else if cur.is_eof() {
            if self.should_add_semicolon() {
                let semi = match self.tokens.last() {
                    Some(t) => Token::new(Semicolon(true), t.line, t.column + 1),
                    None => Token::default(),
                };
                self.tokens.push(semi);
            }
            return Ok(Token::new(Eof, cur.line, cur.column));
        }

        // Ignore whitespace
        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Single line comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Keywords, types, and identifiers
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "fn" => Fn,
                "let" => Let,
                "for" => For,
                "if" => If,
                "else" => Else,
                "extern" => Extern,
                "true" => Bool(true),
                "false" => Bool(false),
                "struct" => Struct,
                "module" => Module,
                _ => Ident(identifier),
            };

            return Ok(Token::new(tt, cur.line, cur.column));
        }

        // Literal numbers
        if cur.value.is_ascii_digit() {
            let mut n = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    n.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return Ok(Token::new(Num(n), cur.line, cur.column));
        }

        // Literal char
        if cur == '\'' {
            let mut ch = String::new();
            let next = self.stream.next().unwrap_or_else(|| unreachable!("lexed None when looking for char"));

            match next.value {
                // Control characters
                '\\' => {
                    if let Some(next) = self.stream.next() {
                        match next.value {
                            'n' => ch = String::from("\n"),
                            't' => ch = String::from("\t"),
                            '\'' => ch = String::from("'"),
                            c => {
                                return Err(LexError::from((
                                    format!("Invalid character control sequence: `\\{}`", c),
                                    next,
                                )))
                            },
                        }
                    }
                },
                // EOF
                '\0' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`, got `EOF`".to_string(),
                        cur,
                    )));
                },
                '\'' => return Err(LexError::from(("Character literal can't be empty".to_string(), cur))),

                // Everything else
                c => ch = String::from(c),
            }

            // Check for closing '\''
            let last =
                self.stream.next().unwrap_or_else(|| unreachable!("fatal: lexed None when looking for `'`"));
            match last.value {
                '\'' => (),
                '\0' | '\n' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`".to_string(),
                        last,
                    )));
                },
                _ => {
                    return Err(LexError::from((
                        format!("Invalid character sequence: `'{}{}'`", ch, last.value),
                        last,
                    )));
                },
            }

            return Ok(Token::new(Char(ch), cur.line, cur.column));
        }

        // Multi-character operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Eq), cur.line, cur.column));
                },
                '!' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::NotEq), cur.line, cur.column));
                },
                '>' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::GtEq), cur.line, cur.column));
                },
                '<' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::LtEq), cur.line, cur.column));
                },
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::And), cur.line, cur.column));
                },
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Or), cur.line, cur.column));
                },
                '+' if next == &'+' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Inc), cur.line, cur.column));
                },
                '+' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::AddEq), cur.line, cur.column));
                },
                '-' if next == &'-' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Dec), cur.line, cur.column));
                },
                '-' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::SubEq), cur.line, cur.column));
                },
                '-' if next == &'>' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::RetType), cur.line, cur.column));
                },
                '*' if next == &'*' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Pow), cur.line, cur.column));
                },
                '*' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::MulEq), cur.line, cur.column));
                },
                '/' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::DivEq), cur.line, cur.column));
                },
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => Op(Operator::Add),
            '-' => Op(Operator::Sub),
            '*' => Op(Operator::Mul),
            '/' => Op(Operator::Div),
            '>' => Op(Operator::Gt),
            '<' => Op(Operator::Lt),
            '!' => Op(Operator::Not),
            '=' => Op(Operator::Assign),
            '&' => Op(Operator::BitAnd),
            '^' => Op(Operator::BitXor),
            '|' => Op(Operator::BitOr),
            '.' => Dot,
            '}' => CloseBrace,
            ']' => CloseBracket,
            ')' => CloseParen,
            ':' => Colon,
            ',' => Comma,
            '{' => OpenBrace,
            '[' => OpenBracket,
            '(' => OpenParen,
            ';' => Semicolon(false),
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            },
        };

        Ok(Token::new(tt, cur.line, cur.column))
    }

    // Add a semicolon for these tokens
    fn should_add_semicolon(&self) -> bool {
        use TokenType::*;

        if let Some(t) = self.tokens.last() {
            matches!(
                t.tt,
                Bool(_)
                    | Char(_)
                    | CloseBrace
                    | CloseParen
                    | CloseBracket
                    | Ident(_)
                    | Num(_)
                    | Op(Operator::Inc)
                    | Op(Operator::Dec)
            )
        } else {
            false
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, Clone, Copy, PartialEq)]
struct ContextElement<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextElement<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextElement { value, line: line + 1, column: column + 1 }
    }
}

impl ContextElement<char> {
    // Will cause lexing to stop if there's a null byte in the file
    fn is_eof(&self) -> bool {
        self.value == 0 as char
    }
}

impl PartialEq<char> for ContextElement<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextElement<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let opt = self.lines.get(self.line);
        let line = match opt {
            Some(l) => l,
            None => return Some(ContextElement::new(0 as char, self.line, self.column - 1)),
        };
        let cc =
            line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column)).or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc.or_else(|| Some(ContextElement::new(0 as char, self.line, self.column - 1)))
    }
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Lexing error: {} at {}:{}", self.message, self.line, self.column)
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextElement<T>)> for LexError {
    fn from((msg, cp): (String, ContextElement<T>)) -> Self {
        LexError { message: msg, line: cp.line, column: cp.column }
    }
}
