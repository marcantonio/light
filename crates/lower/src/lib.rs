use common::symbol_table::Symbolic;
use common::{Literal, Operator, Prototype, Symbol, SymbolTable, Type};
pub use hir::Hir;
use parse::ast::{self, Ast, VisitableNode, Visitor};

pub mod hir;
mod macros;
#[cfg(test)]
mod tests;

// Performs the following:
// - builds the HIR
// - desugars x += 1 to x = x + 1
// - cooks function names in the AST and symbol table
// - tracks scope (needed?)
// - initializes uninitialized variables
// - drops field information from structs
// - inserts let statements to support field/method chaining

pub struct Lower<'a> {
    symbol_table: &'a mut SymbolTable<Symbol>,
    struct_methods: Vec<hir::Node>,
    import_symbols: Vec<Symbol>,
    module: String,
}

impl<'a> Lower<'a> {
    pub fn new(module: &str, symbol_table: &'a mut SymbolTable<Symbol>) -> Self {
        Lower { symbol_table, struct_methods: vec![], import_symbols: vec![], module: module.to_owned() }
    }

    pub fn walk(mut self, ast: Ast<ast::Node>) -> Result<Hir<hir::Node>, String> {
        let mut hir = Hir::new();
        let nodes = ast
            .into_nodes()
            .into_iter()
            .map(|node| node.accept(&mut self))
            .collect::<Result<Vec<_>, String>>()?;

        // Add globals nodes to the right place in the HIR. Make sure struct methods live
        // at the top of the tree after lowering XXX: Why?
        nodes.into_iter().chain(self.struct_methods).for_each(|node| match node.kind {
            hir::node::Kind::Struct { .. } => hir.add_struct(node),
            hir::node::Kind::Fn { ref proto, .. } => {
                hir.add_prototype(proto.clone());
                hir.add_function(node);
            },
            _ => unreachable!("invalid node kind at global level"),
        });

        // XXX: abstract
        // XXX: this could maybe move to visit_call unless structs block it
        for symbol in self.import_symbols {
            hir.add_prototype(Prototype::from(symbol))
        }

        Ok(hir)
    }

    fn lower_lit_array(&mut self, lit: Literal<ast::Node>) -> Result<Literal<hir::Node>, String> {
        // Extract the elements vec and the type of the array elements.
        let (elements, ty) = match lit {
            Literal::Array { elements, inner_ty } => (elements, inner_ty),
            _ => unreachable!("expected array literal"),
        };

        // Rewrap every element
        let mut chkd_elements = Vec::with_capacity(elements.len());
        for el in elements {
            chkd_elements.push(self.visit_node(el)?);
        }

        // Rebuild the literal and return the type
        Ok(Literal::Array { elements: chkd_elements, inner_ty: ty })
    }

    // Helper for variable initializations
    fn lower_var_init(
        &mut self, name: &str, init: Option<&ast::Node>, antn: &Type,
    ) -> Result<hir::Node, String> {
        if let Some(init) = init {
            self.visit_node(init.clone())
        } else {
            self.init_null(name, antn)
        }
    }

    fn init_null(&mut self, name: &str, antn: &Type) -> Result<hir::Node, String> {
        use Type::*;

        Ok(match antn {
            Int8 => init_literal!(Int8, 0),
            Int16 => init_literal!(Int16, 0),
            Int32 => init_literal!(Int32, 0),
            Int64 => init_literal!(Int64, 0),
            UInt8 => init_literal!(UInt8, 0),
            UInt16 => init_literal!(UInt16, 0),
            UInt32 => init_literal!(UInt32, 0),
            UInt64 => init_literal!(UInt64, 0),
            Float => init_literal!(Float, 0.0),
            Double => init_literal!(Double, 0.0),
            Char => init_literal!(Char, 0),
            Bool => init_literal!(Bool, false),
            SArray(ty, len) => hir::Node::new_lit(
                Literal::Array { elements: Vec::with_capacity(*len), inner_ty: Some(*ty.clone()) },
                Type::SArray(Box::new(*ty.clone()), *len),
            ),
            Comp(name) => {
                let sym = self
                    .symbol_table
                    .resolve_symbol(name, &self.module)
                    .cloned()
                    .unwrap_or_else(|| unreachable!("missing symbol for `{}` in `init_null()`", name));
                let initializers = if let Some(fields) = sym.fields() {
                    fields
                        .iter()
                        .map(|(n, a)| self.init_null(n, &(*a).into()))
                        .collect::<Result<Vec<_>, String>>()?
                } else {
                    vec![]
                };
                hir::Node::new_lit(Literal::Comp(initializers), Type::Comp(name.to_owned()))
            },
            Void => unreachable!("void type for `{}` variable initialization annotation", name),
        })
    }
}

impl<'a> ast::Visitor for Lower<'a> {
    type AstNode = ast::Node;
    type Result = Result<hir::Node, String>;

    fn visit_node(&mut self, node: Self::AstNode) -> Self::Result {
        node.accept(self)
    }

    fn visit_for(
        &mut self, start_name: String, start_antn: Type, start_expr: Option<ast::Node>, cond_expr: ast::Node,
        step_expr: ast::Node, body: ast::Node,
    ) -> Self::Result {
        // Insert start var
        self.symbol_table.enter_scope();
        self.symbol_table.insert(Symbol::new_var(&start_name, &start_antn, &self.module));

        let start_expr = self.lower_var_init(&start_name, start_expr.as_ref(), &start_antn)?;
        let cond_expr = self.visit_node(cond_expr)?;
        let step_expr = self.visit_node(step_expr)?;
        let body = self.visit_node(body)?;

        self.symbol_table.leave_scope();

        Ok(hir::Node::new_for(start_name, start_antn, Some(start_expr), cond_expr, step_expr, body))
    }

    fn visit_let(&mut self, name: String, antn: Type, init: Option<ast::Node>) -> Self::Result {
        self.symbol_table.insert(Symbol::new_var(&name, &antn, &self.module));
        let init_node = self.lower_var_init(&name, init.as_ref(), &antn)?;
        Ok(hir::Node::new_let(name, antn, Some(init_node)))
    }

    fn visit_fn(&mut self, proto: Prototype, body: Option<ast::Node>) -> Self::Result {
        let mut proto = proto;
        // Insert a duplicate of the symbol. The new one will have the lowered name. Use
        // updated name in the HIR. Skip for externs.
        let sym = self
            .symbol_table
            .get(proto.name())
            .cloned()
            .unwrap_or_else(|| unreachable!("missing symbol in `visit_fn()` for `{}`", proto.name()));

        if !sym.is_extern() {
            // For use in the HIR
            proto.set_name(sym.name().to_owned());
            // Updates the map key name
            self.symbol_table.insert(sym);
        }

        // This creates an interstitial scope for the arguments in the function definition
        // because lower_block() will also create a new scope. Shouldn't be a practical
        // issue.
        self.symbol_table.enter_scope();

        for arg in proto.args() {
            self.symbol_table.insert(Symbol::new_var(&arg.0, &arg.1, &self.module));
        }

        let body_node = body.map(|e| self.visit_node(e));

        self.symbol_table.leave_scope();

        Ok(hir::Node::new_fn(proto, body_node.transpose()?))
    }

    fn visit_struct(
        &mut self, name: String, fields: Vec<ast::Node>, methods: Vec<ast::Node>,
    ) -> Self::Result {
        let field_tys = fields
            .into_iter()
            .map(|n| match n.kind {
                ast::node::Kind::Let { antn, .. } => antn,
                _ => unreachable!("invalid node type in struct fields"),
            })
            .collect();

        // Save the methods separately to pop them up to the top of the HIR later
        let mut lowered_methods = methods
            .into_iter()
            .map(|n| {
                let mut node = self.visit_node(n)?;

                // Insert `self` into the method's args list.
                match node.kind {
                    hir::node::Kind::Fn { ref mut proto, .. } => {
                        let mut args = vec![(String::from("self"), Type::Comp(name.clone()))];
                        args.append(&mut proto.args().to_vec());
                        proto.set_args(args);
                    },
                    _ => unreachable!("non-function node in struct methods"),
                }
                Ok(node)
            })
            .collect::<Result<Vec<_>, String>>()?;
        self.struct_methods.append(&mut lowered_methods);

        Ok(hir::Node::new_struct(name, field_tys))
    }

    fn visit_lit(&mut self, value: Literal<ast::Node>, ty: Option<Type>) -> Self::Result {
        use Literal::*;

        // Rewrapping primitives is annoying. Remove if we dump ast::Node -> hir::Node
        let lit = match value {
            Int8(l) => Int8(l),
            Int16(l) => Int16(l),
            Int32(l) => Int32(l),
            Int64(l) => Int64(l),
            UInt8(l) => UInt8(l),
            UInt16(l) => UInt16(l),
            UInt32(l) => UInt32(l),
            UInt64(l) => UInt64(l),
            Float(l) => Float(l),
            Double(l) => Double(l),
            Bool(l) => Bool(l),
            Char(l) => Char(l),
            Array { .. } => self.lower_lit_array(value)?,
            Comp(_) => todo!(),
        };
        Ok(hir::Node::new_lit(lit, ty.unwrap_or_default()))
    }

    fn visit_ident(&mut self, name: String, ty: Option<Type>) -> Self::Result {
        Ok(hir::Node::new_ident(name, ty.unwrap_or_default()))
    }

    // Lower `x += 1` to `x = x + 1`
    fn visit_binop(
        &mut self, op: Operator, lhs: ast::Node, rhs: ast::Node, ty: Option<Type>,
    ) -> Self::Result {
        use Operator::*;

        let lowered_lhs = self.visit_node(lhs)?;

        let ty = ty.unwrap_or_default();

        let top_op;
        let rhs = match op {
            AddEq => {
                top_op = Assign;
                hir::Node::new_binop(Add, lowered_lhs.clone(), self.visit_node(rhs)?, ty.clone())
            },
            SubEq => {
                top_op = Assign;
                hir::Node::new_binop(Sub, lowered_lhs.clone(), self.visit_node(rhs)?, ty.clone())
            },
            MulEq => {
                top_op = Assign;
                hir::Node::new_binop(Mul, lowered_lhs.clone(), self.visit_node(rhs)?, ty.clone())
            },
            DivEq => {
                top_op = Assign;
                hir::Node::new_binop(Div, lowered_lhs.clone(), self.visit_node(rhs)?, ty.clone())
            },
            _ => {
                top_op = op;
                self.visit_node(rhs)?
            },
        };

        Ok(hir::Node::new_binop(top_op, lowered_lhs, rhs, ty))
    }

    fn visit_unop(&mut self, op: Operator, rhs: ast::Node, ty: Option<Type>) -> Self::Result {
        Ok(hir::Node::new_unop(op, self.visit_node(rhs)?, ty.unwrap_or_default()))
    }

    fn visit_call(&mut self, name: String, args: Vec<ast::Node>, ty: Option<Type>) -> Self::Result {
        let sym = self
            .symbol_table
            .get(&name)
            .unwrap_or_else(|| unreachable!("missing symbol in `visit_call()` for `{}`", name));

        // Update the HIR with the lowered name if it hasn't been done already and it's
        // not an extern call
        let lowered_name = match sym.name() {
            sym_name if !sym.is_extern() && sym_name != name => sym_name.to_owned(),
            _ => name,
        };

        // Make a list of all imported functions
        if sym.is_import(&self.module) {
            self.import_symbols.push(sym.clone())
        }

        let mut lowered_args = vec![];
        for node in args {
            lowered_args.push(self.visit_node(node)?);
        }
        Ok(hir::Node::new_call(lowered_name, lowered_args, ty.unwrap_or_default()))
    }

    fn visit_cond(
        &mut self, cond_expr: ast::Node, then_block: ast::Node, else_block: Option<ast::Node>,
        ty: Option<Type>,
    ) -> Self::Result {
        Ok(hir::Node::new_cond(
            self.visit_node(cond_expr)?,
            self.visit_node(then_block)?,
            else_block.map(|e| self.visit_node(e)).transpose()?,
            ty.unwrap_or_default(),
        ))
    }

    fn visit_block(&mut self, list: Vec<ast::Node>, ty: Option<Type>) -> Self::Result {
        self.symbol_table.enter_scope();

        let mut lowered_list = vec![];
        for node in list {
            lowered_list.push(self.visit_node(node)?);
        }

        self.symbol_table.leave_scope();

        Ok(hir::Node::new_block(lowered_list, ty.unwrap_or_default()))
    }

    fn visit_index(&mut self, binding: ast::Node, idx: ast::Node, ty: Option<Type>) -> Self::Result {
        Ok(hir::Node::new_index(self.visit_node(binding)?, self.visit_node(idx)?, ty.unwrap_or_default()))
    }

    fn visit_fselector(&mut self, comp: ast::Node, field: String, ty: Option<Type>) -> Self::Result {
        let mut lowered_comp = self.visit_node(comp)?;

        let comp_name = match lowered_comp.ty() {
            Type::Comp(name) => name.to_owned(),
            _ => unreachable!("unexpected type for for field selector target in lower"),
        };

        // If the composite isn't an ident or a let, wrap it in a new let stmt
        if let hir::node::Kind::Call { ty, .. } = lowered_comp.clone().kind {
            lowered_comp = hir::Node::new_let(self.symbol_table.uniq_ident(None), ty, Some(lowered_comp))
        }

        // Find the symbol and extract the index that corresponds to `field`
        let comp_sym = self
            .symbol_table
            .get(&comp_name)
            .unwrap_or_else(|| unreachable!("missing symbol table entry for composite: `{}`", comp_name));
        let idx = comp_sym
            .fields()
            .unwrap_or_default()
            .into_iter()
            .enumerate()
            .find(|(_, f)| f.0 == field)
            .map(|(i, _)| i)
            .unwrap_or_else(|| unreachable!("composite `{}` has no field: `{}`", comp_name, field))
            .try_into()
            .map_err(|err| format!("failed to convert composite index: `{}`", err))?;

        Ok(hir::Node::new_fselector(lowered_comp, idx, ty.unwrap()))
    }

    fn visit_mselector(
        &mut self, comp: ast::Node, name: String, args: Vec<ast::Node>, ty: Option<Type>,
    ) -> Self::Result {
        let lowered_comp = self.visit_node(comp)?;
        let lowered_call = self.visit_call(name, args, ty)?;
        match lowered_call.kind {
            hir::node::Kind::Call { name, mut args, ty } => {
                // xxx: add lowered_comp as self here
                args.insert(0, lowered_comp);
                Ok(hir::Node::new_call(name, args, ty))
            },
            _ => unreachable!("unknown node kind in `visit_mselector()`"),
        }
    }
}
