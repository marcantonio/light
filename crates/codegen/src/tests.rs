use super::*;
use lex::Lex;
use lower::Lower;
use parse::Parse;
use tych::Tych;

macro_rules! run_insta {
    ($prefix:expr, $tests:expr) => {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for test in $tests {
                // Unoptimized code
                let tokens = Lex::new(test[1]).scan().unwrap();
                let mut parser = Parse::new(&tokens);
                let ast = parser.parse().unwrap();
                let mut symbol_table = SymbolTable::new();
                parser.merge_symbols(&mut symbol_table).unwrap();
                let typed_ast = Tych::new(&mut symbol_table).walk(ast).unwrap();
                let hir = Lower::new(vec![], &mut symbol_table).walk(typed_ast).unwrap();
                let args = CliArgs::new();
                let res = Codegen::run(hir, "main", symbol_table, PathBuf::new(), &args, true)
                    .expect("codegen error").as_ir_string();

                // Optimized code
                let tokens = Lex::new(test[1]).scan().unwrap();
                let mut parser = Parse::new(&tokens);
                let ast = parser.parse().unwrap();
                let mut symbol_table = SymbolTable::new();
                parser.merge_symbols(&mut symbol_table).unwrap();
                let typed_ast = Tych::new(&mut symbol_table).walk(ast).unwrap();
                let hir = Lower::new(vec![], &mut symbol_table).walk(typed_ast).unwrap();
                let mut args = CliArgs::new();
                args.opt_level = 1;
                let res_opt = Codegen::run(hir, "main", symbol_table, PathBuf::new(), &args, true)
                    .expect("codegen error").as_ir_string();

                insta::assert_yaml_snapshot!(format!("{}_{}", $prefix, test[0]), (test[1], res, res_opt));
            }
        })
    };
}

#[test]
fn test_block() {
    let tests = [[
        "basic",
        r#"
fn main() {
    {
        10 + 2
        1.0
        5
    }
}
"#,
    ]];
    run_insta!("block", tests);
}

#[test]
fn test_for() {
    let tests = [[
        "basic",
        r#"
extern fn putchar(x: int) -> int
fn main() {
    let x: int = 3
    for y: int = 0; y < x; 1 {
        putchar(46)
    }
    putchar(10)
}
"#,
    ]];
    run_insta!("for", tests);
}

#[test]
fn test_cond() {
    let tests = [
        [
            "cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    }
}
"#,
        ],
        [
            "cond_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    }
}
"#,
        ],
        [
            "if_else_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_cond_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 8 {
        7
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else if plus_one(1) == 2 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_cond_2_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 2 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_else_conds_fail",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 3 {
        2
    } else {
        1
    }
}
"#,
        ],
        [
            "if_else_if_cond_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 7 {
        7
    } else if plus_one(1) == 1 {
        2
    }
}
"#,
        ],
        [
            "if_else_if_cond_2_success",
            r#"
fn plus_one(x: int) -> int { x + 1 }
fn main() {
    if plus_one(6) == 5 {
        7
    } else if plus_one(1) == 2 {
        2
    }
}
"#,
        ],
        [
            "true_lit",
            r#"
fn main() {
    if true {
        7
    } else {
        2
    }
}
"#,
        ],
        [
            "false_lit",
            r#"
fn main() {
    if false {
        7
    } else {
        2
    }
}
"#,
        ],
    ];
    run_insta!("cond", tests);
}

#[test]
fn test_let() {
    let tests = [
        [
            "basic",
            r#"
        fn main() {
    let x: int = 1
}
"#,
        ],
        [
            "float_default",
            r#"
fn main() {
    let x: float
}
"#,
        ],
        [
            "bool",
            r#"
fn main() {
    let x: bool
}
"#,
        ],
    ];
    run_insta!("let", tests);
}

#[test]
fn test_scope() {
    let tests = [
        [
            "basic_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
    }
    b
}
fn main() { foo(3) }
"#,
        ],
        [
            "nested_shadowing",
            r#"
fn foo(a: int) -> int {
    let b: int = 1
    {
        let b: bool = false
        let a: float = {
            let b: float = 1.0
            b
        }
    }
    b
}
fn main() { foo(3) }
"#,
        ],
        [
            "for",
            r#"
fn foo() {
    let x: float = 1.0
    for x: int = 1; x < 10; 1 {
        x
    }
    x
}
fn main() { foo() }
"#,
        ],
        [
            "if",
            r#"
fn foo() {
    let x: float = 1.0
    if x < 2.0 {
        let y: int = 2
        x
    }
    x
}
fn main() { foo() }
"#,
        ],
        [
            "if_else",
            r#"
fn foo() {
    let x: float = 1.0
    if x < 2.0 {
        let y: int = 4 & 3 ^ 1
        y
    } else {
        let y: int = -2
        y
    }
}
fn main() { foo() }
"#,
        ],
    ];
    run_insta!("scope", tests);
}

#[test]
fn test_unop() {
    let tests = [[
        "basic",
        r#"
fn test(x: int) -> int {
    x + 1
}

fn main() {
    test(-40)
}
"#,
    ]];
    run_insta!("unop", tests);
}

#[test]
fn test_array() {
    let tests = [[
        "basic",
        r#"
fn main() {
    let a: [int; 3] = [1, 2, 3]
    let b: [int; 3]
    a[1] = 7
    a[1]
}
"#,
    ]];
    run_insta!("array", tests);
}

#[test]
fn test_order() {
    let tests = [
        [
            "a",
            r#"
fn foo() {}
fn main() {
    foo()
}
"#,
        ],
        [
            "b",
            r#"
fn main() {
    foo()
}
fn foo() {}
"#,
        ],
    ];
    run_insta!("order", tests);
}

#[test]
fn test_struct() {
    let tests = [
        [
            "basic",
            r#"
struct Foo {
    let a: int
    let b: bool
    fn c(d: int) -> int { self.a + d }
}
fn main() {
    let x: Foo
    x.a
    x.c(2)
}
"#,
        ],
        [
            "selector",
            r#"
fn returnStruct() -> Foo {
    let a: Foo
    a.a = 1
    a
}
fn main() {
    let x: Foo
    x.a
    let b: Bar
    b.foo.a = returnStruct().a
    b.foo.b()
    b.foo.a
}
struct Foo {
    let a: int
    fn b() {}
}
struct Bar {
    let foo: Foo
}
"#,
        ],
    ];
    run_insta!("struct", tests);
}
