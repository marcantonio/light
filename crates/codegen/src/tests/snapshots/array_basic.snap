---
source: crates/codegen/src/tests.rs
assertion_line: 368
expression: "(test[1], res, res_opt)"
---
- "\nfn main() {\n    let a: [int; 3] = [1, 2, 3]\n    let b: [int; 3]\n    a[1] = 7\n    a[1]\n}\n"
- "; ModuleID = 'main'\nsource_filename = \"main\"\n\ndefine void @main() {\nentry:\n  %b = alloca [3 x i32], align 4\n  %a = alloca [3 x i32], align 4\n  store [3 x i32] [i32 1, i32 2, i32 3], [3 x i32]* %a, align 4\n  store [3 x i32] zeroinitializer, [3 x i32]* %b, align 4\n  %array.index.gep = getelementptr inbounds [3 x i32], [3 x i32]* %a, i32 0, i32 1\n  %index.a = load i32, i32* %array.index.gep, align 4\n  %array.index.gep1 = getelementptr inbounds [3 x i32], [3 x i32]* %a, i32 0, i32 1\n  store i32 7, i32* %array.index.gep1, align 4\n  %array.index.gep2 = getelementptr inbounds [3 x i32], [3 x i32]* %a, i32 0, i32 1\n  %index.a3 = load i32, i32* %array.index.gep2, align 4\n  ret void\n}\n"
- "; ModuleID = 'main'\nsource_filename = \"main\"\n\ndefine void @main() {\nentry:\n  ret void\n}\n"

